File: sanghyun/programmers-2/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String,String> input = new LinkedHashMap<>();
        Queue<String[]> queue = new LinkedList<>();
        int cnt = 0;
        for (int i = 0; i < record.length; i++) {
            String type = record[i].split(" ")[0];
            String uid = record[i].split(" ")[1];
            String[] s = {type,uid};
            if(type.equals("Change")) {
                input.put(uid,record[i].split(" ")[2]);
            }else if(type.equals("Enter")){
                cnt++;
                input.put(uid,record[i].split(" ")[2]);
                queue.add(s);
            }else{
                cnt++;
                queue.add(s);
            }
        }
        String[] answer = new String[cnt];
        for (int i = 0; i < cnt; i++) {
            String type = queue.peek()[0];
            String uid = queue.peek()[1];
            queue.poll();
            if(type.equals("Enter")){
                answer[i] = input.get(uid)+"님이 들어왔습니다.";
            }else{
                answer[i] = input.get(uid)+"님이 나갔습니다.";
            }
        }
        return answer;
    }
}

File: sanghoon/programmers-2/오픈채팅방.py
def solution(record):
    answer = []
    user_list = {}

    for rec in record:
        v = rec.split()

        if v[0] in ['Enter', 'Change']:
            user_list[v[1]] = v[2]

    for rec in record:
        v = rec.split()
        if v[0] == 'Enter':
            answer.append(f'{user_list[v[1]]}님이 들어왔습니다.')
        elif v[0] == 'Leave':
            answer.append(f'{user_list[v[1]]}님이 나갔습니다.')

    return answer

record = ["Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"]
solution(record)